### PR TITLE
Translate `_d_newclass` to a template

### PR DESCRIPTION
**Pull request by teodutu**
This PR makes the following changes:
- Add template `_d_newclassT` to `druntime.src.core.lifetime.d`
- Replace lowering of `new C()` to `_d_newclassT!C()`
- Add `lowering` member to `NewExp`. This field stores the above lowering to be used by e2ir.d
- Keep the old `_d_newclass` hook because it's used by `TypeInfo_Class.create()`
- Add dummy `_d_newclassT` hook to tests that redefine the `object` module
- Remove `new MinHeap!(TestType)()` from `fail308.d`. Otherwise the errror was changed and printed the local path to druntime
- Account for the `GC.malloc()` called by the template hook in the `-profile=gc` tests

### CODE REVIEW

**Review by ibuclaw — DISMISSED** [review]

File: compiler/src/dmd/expression.d [hunk 1]
```diff
@@ -3576,6 +3577,8 @@ extern (C++) final class NewExp : Expression
     bool onstack;               // allocate on stack
     bool thrownew;              // this NewExp is the expression of a ThrowStatement
 
+    Expression lowering;        // lowered druntime hook: `_d_newclass`
```
[the remark this answers is not in the document]
Comment (teodutu) [reply compiler/src/dmd/expression.d:3580]: Done. Thanks!

Comment (maxhaton) [inline compiler/src/dmd/expression.d:3580]: Is this necessary? If yes keep it but keep an eye on adding unnecessary fields as dmd's classes are absolutely enormous

**Review by RazvanN7 — COMMENTED** [review]
Why aren't we deleting the old hook?

**Comment by teodutu** [comment]
> Why aren't we deleting the old hook?

Because it's used by [`TypeInfo_Class.create()`](https://github.com/dlang/dmd/blob/67c137a1774976f6539e2f4897a3b153658e303f/druntime/src/object.d#L1690-L1702), which is publicly exported.

**Comment by RazvanN7** [comment]
Can't we just call the template?

**Comment by teodutu** [comment]
> 

Maybe, but for this, we would need the actual type to be available at compile time. We can obtain it from `TypeInfo_Class.name`, but only at run time.

File: compiler/src/dmd/e2ir.d [hunk 2]
```diff
@@ -1083,10 +1083,7 @@ elem* toElem(Expression e, IRState *irs)
                 assert(!(global.params.ehnogc && ne.thrownew),
                     "This should have been rewritten to `_d_newThrowable` in the semantic phase.");
 
-                Symbol *csym = toSymbol(cd);
-                const rtl = RTLSYM.NEWCLASS;
-                ex = el_bin(OPcall,TYnptr,el_var(getRtlsym(rtl)),el_ptr(csym));
-                toTraceGC(irs, ex, ne.loc);
+                ex = toElem(ne.lowering, irs);
```
Comment (ibuclaw) [inline compiler/src/dmd/e2ir.d:1086]: Are previous lowerings going to be retroactively transplanted into a lowering field?

Comment (teodutu) [reply compiler/src/dmd/e2ir.d:1086]: Yes, that's the plan.

File: druntime/.gdb_history [hunk 3]
```diff
@@ -0,0 +1,3 @@
+r
+bt
+q
```
Comment (dkorpel) [inline druntime/.gdb_history:3]: Accidental commit

File: druntime/src/rt/sections.d [hunk 4]
```diff
@@ -100,7 +102,6 @@ bool scanDataSegPrecisely() nothrow @nogc
         case "precise":
             return true;
         default:
-            __gshared err = new Error("DRT invalid scanDataSeg option, must be 'precise' or 'conservative'");
```
Comment (dkorpel) [inline druntime/src/rt/sections.d:103]: > Move err to global scope in rt.sections.d to avoid the frontend lowering

Why is this needed?

**Review by dkorpel — DISMISSED** [review]

File: druntime/.gdb_history (hunk 3 above)
Comment (teodutu) [reply druntime/.gdb_history]: :man_facepalming: I removed it now.

File: druntime/src/rt/sections.d (hunk 4 above)
Comment (teodutu) [reply druntime/src/rt/sections.d]: I was getting an error for calling the hook that calls `GC.malloc()` from this `@nogc` function. Now after rebasing it seems the error is fixed, so I rolled back this change.

**Comment by RazvanN7** [comment]
@thewilsonator why have you removed the `72h no objections->merge` label. Is there anything that you are requesting?

**Comment by thewilsonator** [comment]
https://github.com/dlang/dmd/pull/14837#pullrequestreview-1254083963

**Comment by RazvanN7** [comment]
It has been addressed.

**Comment by RazvanN7** [comment]
@ibuclaw is this ready to go?

**Comment by RazvanN7** [comment]
@thewilsonator @ibuclaw re-adding the `72h no objections -> merge` label

File: druntime/src/core/lifetime.d [hunk 5]
```diff
@@ -2709,3 +2709,108 @@ T _d_newThrowable(T)() @trusted
...
+// Test allocation
+@safe unittest
+{
+    class C { }
+    C c = _d_newclassT!C();
+
+    assert(c !is null);
+}
+
+// Test initialisers
```
Comment (dkorpel) [inline druntime/src/core/lifetime.d:2786]:
```suggestion
// Test initializers
```

**Review by dkorpel — APPROVED** [review]

**Comment by jsatellite** [comment]
This changes how COM classes are managed - the memory will now be allocated by the GC whereas currently `_d_newclass` uses malloc. Looks like the getLinkage trait replaces the test for `ci.m_flags & TypeInfo_Class.ClassFlags.isCOMClass` - but __traits(getLinkage, ...) doesn't return "Windows" even if the class is declared with `extern(Windows)` (is it supposed to?). COM classes should either use malloc, or GC memory and a root added.

**Comment by rainers** [comment]
> This changes how COM classes are managed

Good catch. The linkage reported for a class is derived from its member `classKind` which is C++ for COM classes.

> COM classes should either use malloc, or GC memory and a root added.

IMO GC memory is the right choice, but adding/removing roots should be done in AddRef/Release. See https://issues.dlang.org/show_bug.cgi?id=4092

**Comment by RazvanN7** [comment]
> doesn't return "Windows" even if the class is declared with extern(Windows) (is it supposed to?)

It returns Windows only if compiled for 32-bit.

**Comment by RazvanN7** [comment]
How are we supposed to test if a class is a COM class? See if it implements IUnknown? Is there an easier way?

**Comment by jsatellite** [comment]
Can't you use `T.classinfo.m_flags & TypeInfo_Class.ClassFlags.isCOMClass` like in the un-templated _d_newclass?

**Comment by ibuclaw** [comment]
This PR introduced a regression https://issues.dlang.org/show_bug.cgi?id=23688

File: compiler/src/dmd/expression.h [hunk 6]
```diff
@@ -539,6 +539,8 @@ class NewExp final : public Expression
     bool onstack;               // allocate on stack
     bool thrownew;              // this NewExp is the expression of a ThrowStatement
 
+    Expression lowering;        // lowered druntime hook: `_d_newclass`
```
Comment (RazvanN7) [inline compiler/src/dmd/expression.h:542]: Shouldn't have this been a pointer?

**Comment by ibuclaw** [comment]
This PR caused a regression.

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=110113

I'm going to have to revert this PR if a proper fix doesn't come in a reasonable time-frame given the severity of it.

**Comment by ibuclaw** [comment]
> Does it only trigger with dip1021?

All that dip1021 is doing is making the compiler do a few more GC allocations/freeing by executing this code.

https://github.com/dlang/dmd/blob/master/compiler/src/dmd/ob.d

I think more likely, the trigger is `-lowmem`.  Turning on dip1021 just allows the mesns.

Upstream bug report. https://issues.dlang.org/show_bug.cgi?id=23978
